### PR TITLE
ansible-inventory cli tool

### LINKED ISSUE #12092: ansible inventory cli
##### ISSUE TYPE
Feature Idea

##### COMPONENT NAME
core

##### ANSIBLE VERSION
2.1

##### SUMMARY
This would incorporate the existing 3rd party:
- ansible-inventory: shows hosts and associated vars https://github.com/srvg/ansible-plugins/blob/devel/bin/ansible-inventory
- ansible-inventory-grapher: shows graph of inventory group and host relationships https://github.com/willthames/ansible-inventory-grapher
- ansible-inventory enable <dynamic inventory> : feature to help users setup and manage dynamic inventory sources included with ansible (contrib/inventory).

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY

tool to inspect inventory as ansible sees it (can act as inventory script)

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME

inventory
cli

##### ANSIBLE VERSION

```
2.4
```

##### ADDITIONAL INFORMATION

TODO?: add inventory script enable/disable/config
 make compatible with #23001

fixes #12092

### CODE REVIEW

File: lib/ansible/cli/inventory.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+# ansible-vault is a script that encrypts/decrypts YAML files. See
```
Comment (alikins) [inline lib/ansible/cli/inventory.py]:
```
# ansible-vault is a script that encrypts/decrypts YAML files. See
# http://docs.ansible.com/playbooks_vault.html for more details.
```

That is true, but probably not needed here ;->

Comment (bcoca) [reply lib/ansible/cli/inventory.py]: my subliminal publicity for vault is foiled!

**Comment by bcoca** [comment]
cc @willthames @srvg

**Comment by willthames** [comment]
It seems to work the other way up to ansible-inventory-grapher

For example, if I want to see the groups that a host comes from, I can just run it against that host.

To get the hiearchy for a single host using ansible-inventory, you have to start from all.

If this worked against a single host, and displayed the whole hierarchy above the host, it would be more useful (ansible-inventory-grapher can take a single host or group and goes up to the root to display the whole tree).

It's going to be difficult to see unexpected sibling relationships in a terminal, which is why ansible-inventory-grapher outputs to dot format by default

File: lib/ansible/cli/inventory.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+        results = {}
+        top = self.inventory.get_group('all')
+        results = format_group(top)
+        results['_meta'] = {'hostvars': {}}
+        hosts = self.inventory.get_hosts()
+        for host in hosts:
+            results['_meta']['hostvars'][host.name] = host.get_vars()
+
+            self._remove_internal(results['meta']['hostvars'][host.name])
```
Comment (srgvg) [inline lib/ansible/cli/inventory.py]: The leading underscore in the '_meta' key is missing here.

**Comment by bcoca** [comment]
@willthames it is missing a few more features I would like to add, that is why I cc'd you guys on this.
also, updated to output cleaner JSON and 'usable as inventory' YAML now.

current functions: 
* can be used as inventory script (needed once we start moving scripts into plugins https://github.com/ansible/proposals/issues/41), `--list`, `--host` provide this
* can be used to generate/migrate to YAML format (`--list -y`)
* terminal visible hier of inventory (`--graph`)

I would like to add:
* host centric view of relationships (`--graph-host`?)
* var resolution tracking 
* actual image graphs (`--img`?)
* inventory script/plugin management

**Comment by jctanner** [comment]
shipit yesterday

**Comment by abadger** [comment]
The code seems fine but this brings up a couple high level questions:
*  do we want to have a "complete set of tools" inside of ansible/ansible?  Or do we want to concentrate on core features and let extra functionality be driven by external projects?
* Should we enhance one of the existing tools instead of writing our own thing?
  * If we enhance one of the existing tools,  would we want to eventually merge it into ansible/ansible or just contribute to the external project?

**Comment by cchurch** [comment]
@bcoca It doesn't appear to include any variables defined in YAML files under `host_vars` or `group_vars`.

I created a standalone version that works as a custom inventory script within Tower using Ansible 2.0-2.2 (https://gist.github.com/cchurch/3b6027d3816c0ea3efe42b55ba924c09), and it does output group variables when using 2.0, but not 2.1 or later.

**Comment by bcoca** [comment]
@cchurch fixed, just note that w/o a playbook, these will always be 'just inventory adjacent' ones.

**Comment by mattclay** [comment]
For the indentation issue, `pep8` is expecting this:

```
INTERNAL_VARS = frozenset([
    'ansible_version',
    'ansible_playbook_python',
    'inventory_dir',
    'inventory_file',
    'inventory_hostname',
    'inventory_hostname_short',
    'groups',
    'group_names',
    'omit',
    'playbook_dir',
])
```

**Comment by bcoca** [comment]
@mattclay, can we eliminate that check, it really adds nothing to maintainability and it def is annoying

**Comment by willthames** [comment]
The other way to do it is to avoid the hanging indent

```
INTERNAL_VARS = frozenset(['ansible_version',
                           'ansible_playbook_python',
                           'inventory_dir',
                           ...
])
```

I think aspiring to pep8 is reasonable.

**Comment by bcoca** [comment]
I'ld like someone to explain to me why the following is 'unmaintainable' compared to the 'solutions'
```
INTERNAL_VARS = frozenset([  
                           'ansible_version',
                           'ansible_playbook_python',
                           'inventory_dir',
                           'inventory_file',
                           'inventory_hostname',
                           'inventory_hostname_short',
                           'groups',
                           'group_names',
                           'omit',
                           'playbook_dir',
                         ])
```

**Comment by willthames** [comment]
@bcoca, I don't know, maybe read PEP8. But basically it's not either of the two 'recommended' ways of doing it - it's its own special case. I just don't know why you'd want to indent to that indentation level rather than either of the two 'normal' ways.

It's not that it's unmaintainable, clearly it's maintainable, it's that it's inconsistent with python whitespace best practices.

The amount of horrible looking stuff I see reviewing the codebase (and this is a long way from that), the more I want to be able to say 'this does not meet guidelines' - it's easier to say 'if it meets PEP8 it's good' than allow special cases.

**Comment by bcoca** [comment]
@willthames there is a line between keeping clean code and counterproductive styling, I think this crosses that line and depending where you are in the code pushing the block right or left will make it clearer and more readable, but this rule actually forces you to do otherwise. 

we happily ignore E127, yet keep E126 which is just as counterproductive in my view.

**Comment by mattclay** [comment]
@bcoca If you think relaxing code-style checks warrants further discussion, include it on the TWG agenda for next week. I'd like to see us be able to put this issue to rest, one way or another.

**Comment by sivel** [comment]
> we happily ignore E127, yet keep E126 which is just as counterproductive in my view

We no longer ignore either of those in new files.  The only ignore for new files is now E402.

Also, I think "unmaintainable" is the wrong word.  Style guidelines are about readability, and my personal opinion, is what you have makes no sense from an indentation perspective.  It seems randomly indented, and not aligned with any of the code.  It's merely just indented.

**Comment by sivel** [comment]
The pep8 failures against the current state of the pep8 configuration are:

```
ERROR: Found 15 pep8 issue(s) which need to be resolved:
ERROR: lib/ansible/cli/inventory.py:37:25: E126 continuation line over-indented for hanging indent
ERROR: lib/ansible/cli/inventory.py:47:25: E123 closing bracket does not match indentation of opening bracket's line
ERROR: lib/ansible/cli/inventory.py:49:1: E302 expected 2 blank lines, found 1
ERROR: lib/ansible/cli/inventory.py:68:77: E231 missing whitespace after ','
ERROR: lib/ansible/cli/inventory.py:76:33: E127 continuation line over-indented for visual indent
ERROR: lib/ansible/cli/inventory.py:78:33: E127 continuation line over-indented for visual indent
ERROR: lib/ansible/cli/inventory.py:103:5: E303 too many blank lines (2)
ERROR: lib/ansible/cli/inventory.py:136:13: E265 block comment should start with '# '
ERROR: lib/ansible/cli/inventory.py:156:5: E303 too many blank lines (2)
ERROR: lib/ansible/cli/inventory.py:169:5: E303 too many blank lines (2)
ERROR: lib/ansible/cli/inventory.py:185:22: E231 missing whitespace after ','
ERROR: lib/ansible/cli/inventory.py:186:80: E226 missing whitespace around arithmetic operator
ERROR: lib/ansible/cli/inventory.py:219:5: E303 too many blank lines (2)
ERROR: lib/ansible/cli/inventory.py:268:43: E261 at least two spaces before inline comment
ERROR: lib/ansible/cli/inventory.py:278:9: E303 too many blank lines (2)
ERROR: The 1 sanity test(s) listed below (out of 1) failed. See error output above for details.
```

**Comment by bcoca** [comment]
> We no longer ignore either of those in new files. The only ignore for new files is now E402.


when was this decided? we've already had a 'retroactive proposal' this was not brought up in any recent core meeting i can find.

the only thing i can find is that at one point we decided for NEW MODULES that we would start enforcing more of the PEP standard.

**Comment by bcoca** [comment]
@mattclay it lines up with the definition of the structure itself, not sure how that is not relevant, the first 'fix' showes it 'underindendted' to the structure itself, the 2nd is almost the same as current but with the first element on the same line as the start of the structure.

**Comment by sivel** [comment]
The retroactive proposal covers the info for new files and existing files.

As far as removing the check I'm not sure it makes sense to do so.  Consistent indentation is a benefit, not checking it means we lose standardization.  Standardization means it's easier to read and people know what to expect.  

We can debate or discuss this off this issue. I don't want this distract from the purpose of this issue, and clutter the comments, unless you would like this dicussion to continue here.

**Comment by cchurch** [comment]
@bcoca: Inventory adjacent ones are all I would expect. It appears that any variables defined in files under `group_vars` are not listed under the group itself, but are only included under the resulting host variables for each host.

**Comment by bcoca** [comment]
@cchurch  it consumes the inventory as Ansible  playbooks see it, groups at that point don't exist as entities and cannot have vars themselves.

**Comment by sivel** [comment]
fwiw, I had written an ini2json.py script that could take in a "ini" inventory file and spit out a dynamic inventory JSON output.  It has some ways to still grab group_vars.  Effectively it worked similarly by recreating the JSON output, off of the `inventory` object.

If you are interested: https://gist.github.com/sivel/3c0745243787b9899486

**Comment by bcoca** [comment]
@sivel the advantage of this tool is that it doesn't matter what the source(s) inventory is, as long as it works for Ansible. This is mostly ensuring forward compatibility once #23001 is merged.

**Comment by sivel** [comment]
I get that. Just offering up previous work in case it was helpful.

**Comment by AlanCoding** [comment]
I'm trying to check out this branch and try it but I'm not having much luck so far. Expected use pattern:

`ansible-inventory -i inventory_file.ini --list`

I did a simple `python setup.py install` in a venv, but I didn't find that the `ansible-inventory` CLI command was installed. Not sure if that's my error or something that hasn't been added to this yet. Next, I tried:

`python lib/ansible/cli/inventory.py -i <repo location>/top_level_file.ini --list`

Where <repo location> is a directory with a clone of [this example repo](https://github.com/AlanCoding/Ansible-inventory-file-examples). But every time I run this, I get no output. My expectation was that it would output JSON representing the inventory to the console. Is there something big I'm missing? Sorry I'm still new to playing with the devel version of Ansible.

**Comment by bcoca** [comment]
It should work from checkout once you `source hacking/env-setup`. `ansible-inventory` is a symlink to `ansible` in bin/, which will invoke the class in ` lib/ansible/cli/inventory.py` correclty.

**Comment by AlanCoding** [comment]
Nice! Thanks, the intended use pattern is working out great now.

I just want to ask about one thing - the host vars are all putting in empty `ansible_facts`. Is this intended? This is the output from the example I was trying earlier.

```json
{
    "_meta": {
        "hostvars": {
            "bar1": {
                "ansible_facts": {}
            }, 
            "bar2": {
                "ansible_facts": {}
            }, 
            "foo1": {
                "ansible_facts": {}
            }, 
            "foo2": {
                "ansible_facts": {}
            }
        }
    }, 
    "all": {
        "children": [
            "barland", 
            "fooland", 
            "ungrouped"
        ]
    }, 
    "barland": {
        "hosts": [
            "bar1", 
            "bar2"
        ]
    }, 
    "fooland": {
        "hosts": [
            "foo1", 
            "foo2"
        ]
    }, 
    "ungrouped": {}
}
```

**Comment by bcoca** [comment]
it should not be 'full', not sure why an empty one is appearing, pushed fix for taht

**Comment by AlanCoding** [comment]
Confirmed that fix eliminated the `ansible_facts` dict. It returns the expected content for that inventory file now.

**Comment by AlanCoding** [comment]
One more test case jumped out at me - [here is the test inventory file](https://github.com/AlanCoding/Ansible-inventory-file-examples/blob/master/test_batch/generic.ini) that I'm running the command `ansible-inventory -i ~/repos/Ansible-inventory-file-examples/test_batch/generic.ini --list` against. [Here is the output](https://gist.github.com/AlanCoding/670108b26c8feb373c3c081043be3b9c) that it yields.

The problem is that it gives `"revision": "3.1.2      # Variable definition"` where I was expecting `"revision": "3.1.2"`, chopping off the comment. The inventory file lines that defined this variable are:

```
[webservers:vars]   # Group variables section
revision=3.1.2      # Variable definition
```

It seems that this only happens on lines where we exclusively defining a variable, but lines like `bar1.example.com    # Comment on a host` or `bar2.example.com foo=bar # Comment on an in-line variable definition` are parsed correctly.

Not sure if this the intended behavior, if I should file it as an unrelated issue, or it's an `ansible-inventory` problem.

**Comment by bcoca** [comment]
ansible-inventory does not change how inventory is parsed, just deals with showing you what 'Ansible sees'. So if this is a bug it is on the inventory parsing side.

**Comment by AlanCoding** [comment]
Thanks, I understand. This might fall under the same category, but I want to note another similar behavior. I define a group variable inside of a `group_vars/` directory, living in the same directory as an inventory file. I have a group (call it `my_group`) defined in the inventory file and a file `group_vars/my_group` that defines a group variable. The variable is included in the group's hosts as `hostvars`, but it is not included as a variable for the group. [Here are the files](https://github.com/AlanCoding/Ansible-inventory-file-examples/tree/master/issues/var_not_saved_to_group) and [here is the output](https://gist.github.com/AlanCoding/a4e984cb6383b8f85c09268d864ddb91).

That might not be a functional difference, because I don't know where (inside a playbook run) it might matter that a group _doesn't_ have a variable when all its hosts _do_. If I can contrive of such a situation I'll raise it as a general issue.

**Comment by bcoca** [comment]
That is how Ansible itself gets vars, never from groups (they don't exist for a play) only hosts.

**Comment by AlanCoding** [comment]
One thing I'm noticing is that when the specified inventory file is not found, it still gives an exit code of 0. I see how this behavior makes sense for the normal use of the `-i` option when running Ansible commands, but it seems likely to cause trouble for someone making use of this command. It's not a huge problem to work around - just wanted to raise the issue.

Here's a repo for maintaining the backport, thanks for your comment.

https://github.com/ansible/ansible-inventory-backport

**Comment by AlanCoding** [comment]
+1

@bcoca what's the status of this feature?

**Comment by bcoca** [comment]
limbo, pushed new inventory structure into 2.4, this needs rework .. .also there is doubt this will be included in main repo

**Comment by flare-ws** [comment]
Interested in this, though I think that save/load methods should be implemented in inventory class itself.

I've wrote some quick-and-dirty code a while ago (Methods and classes were partially wiped off):
https://gist.github.com/flare-ws/1cbafd0e84cb8c495d490c5fe9ef2e86

I used this class as an interface to multiple inventory sources (openstack, docker, ini-files) both in python scripts, playbooks (as a dynamic inventory) and as cli utility (especially for dumping ini file).

Now I want to rewrite this for 2.4 and can use some help, especially in design and review. Honestly, i'm a newbie here :(

**Comment by AlanCoding** [comment]
The code in this PR isn't yet compatible with current devel, because the inventory import needs to be switched over to the InventoryManager, so that's one thing that you would need to update your script, but I don't know the extent of the specifics this update will require.

**Comment by bcoca** [comment]
@AlanCoding no worries, i wrote both, pretty aware of the diffs.

@flare-ws the reason I don't want that code in the inventory itself is to allow for more flexible 'dumps', I.e this implements yaml and json, but there should be no reason we cannot use 'user defined templates' to create other formats like ini, csv, etc.

**Comment by flare-ws** [comment]
@bcoca yep, InventoryManager + plugins fits here. I only worry about api now, will there be a way to populate Inventory() from same script and, maybe, append something? Like this:
```
someinstance = SomeClass()
inv = Inventory.Inventory()
inv.parse(someinstance.some_other_method_returns_whatever(**params))
inv.parse(another_data_source) 
<... call play ...>
```

**Comment by bcoca** [comment]
I dont follow...

**Comment by AlanCoding** [comment]
I checked out the most recent version of this branch, then I did the `hacking/env-setup`. I tried against against a number of my examples, and they all seem to be missing the groups entirely. Hosts will include the group vars, but the only group structure shown is all/ungrouped. Is there something else I should try to debug this situation, or can someone else confirm the behavior? `ansible --version` is correctly telling 2.4.

**Comment by bcoca** [comment]
@AlanCoding  I get the groups:

```
all:
  children:
    a_b:
      hosts:
        nyx:
          ansible_executable: /bin/bash
          ansible_ssh_host: 192.168.0.4
    tests:
      hosts:
        centos7test:
          ansible_connection: ssh
          ansible_ssh_host: 192.168.0.24

   ubuntu:
      hosts:
        ubuntu12test: {}
        ubuntu14test: {}
        ubuntu15.10test:
          ansible_connection: ssh
```

**Comment by AlanCoding** [comment]
I replicated these steps using a container to get it reproducible

https://gist.github.com/AlanCoding/1f0add320ccf6c04ee00b03b4b37af00

I'm wondering if something jumps out to you as obviously wrong here? It will echo the hosts in `_meta`, but no groups, where "fooland" and "barland" were expected. Thanks.

Addition:

I also just tried the `python setup.py install`, and this will install Ansible 2.4, but it won't create the links for CLI use. I'm guessing that still hasn't (or won't) be done, but it's something that I will be needing to figure out soon.

**Comment by bcoca** [comment]
not as much an issue with ansible-inventory as the groups not being made children of `all`

**Comment by AlanCoding** [comment]
If I run a playbook, and loop over `groups` inside of that playbook, the groups that are not children of the "all" group will still be used as items inside of that loop (I checked this with Ansible 2.4, and it still works that way). The previous version of the PR would output the following for the example I was using, and you can get similar results for examples in official docs:

```json
{
    "_meta": {
        "hostvars": {
            "bar1": {}, 
            "bar2": {}, 
            "foo1": {}, 
            "foo2": {}
        }
    }, 
    "all": {
        "children": [
            "barland", 
            "fooland", 
            "ungrouped"
        ]
    }, 
    "barland": {
        "hosts": [
            "bar1", 
            "bar2"
        ]
    }, 
    "fooland": {
        "hosts": [
            "foo1", 
            "foo2"
        ]
    }, 
    "ungrouped": {}
}
```

The groups were present as top-level items next to "all" and "ungrouped". If those are not surfaced in this way, then the output of `ansible-inventory` will not reflect the inventory "as ansible sees it".

File: lib/ansible/cli/inventory.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+
+    def run(self):
+
+        results = None
+
+        super(InventoryCLI, self).run()
+
+        # Initialize needed objects
+        if getattr(self, '_play_prereqs'):
```
Comment (AlanCoding) [inline lib/ansible/cli/inventory.py]: I can get this to work with 2.3 if I do `if getattr(self, '_play_prereqs', None):` instead.

**Comment by bcoca** [comment]
@AlanCoding this just revealed a bug with inventory and ini parser/plugin, 'all' is always supposed to be the 'top group'. partially fixed by #26770. 

The issue was that ansible-inventory thought this assumption to be true and iterates from 'all' following the chain. In any case I'll fix ansilbe-inventory also to 'work with bad versions'

File: lib/ansible/cli/inventory.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/cli/inventory.py]: this does not fix 2.4 though

**Comment by AlanCoding** [comment]
It looks like the branch for this PR needs to be rebased in order for the fix from #26770 to take effect.

Even after I rebase, I noticed that all hosts seem to get listed under "ungrouped", including those that are members of other groups. I would imagine that this is unintended behavior. The data I used to test this can be found at: https://github.com/AlanCoding/ansible-inventory-backport/tree/1d53884b15e93661e4c65c03e54fad0caa21ec18/tests/data/everything

Here is the output of parsing the `inventory.ini` file.

```json
{
    "_meta": {
        "hostvars": {
            "host_1": {
                "groupvar_file": "defined_in_group_vars", 
                "groupvar_inventory": "defined_in_inventory_file", 
                "hostvar_file": "defined_in_host_vars", 
                "hostvar_inventory": "defined_in_inventory_file"
            }, 
            "host_2": {
                "groupvar_file": "defined_in_group_vars", 
                "groupvar_inventory": "defined_in_inventory_file"
            }, 
            "host_ungrouped": {}
        }
    }, 
    "all": {
        "children": [
            "group_1", 
            "ungrouped"
        ]
    }, 
    "group_1": {
        "children": [
            "group_2"
        ], 
        "hosts": [
            "host_1"
        ], 
        "vars": {
            "groupvar_inventory": "defined_in_inventory_file"
        }
    }, 
    "group_2": {
        "hosts": [
            "host_2"
        ]
    }, 
    "ungrouped": {
        "hosts": [
            "host_1", 
            "host_2", 
            "host_ungrouped"
        ]
    }
}
```

You can see that both `host_1` and `host_2` are listed in their own groups as well as the ungrouped group.

This would not be a problem for my use case of ansible-inventory, but I thought you would like to know.

**Pull request merged**